Tests, CI, Lint

aminya opened the pull request:
- Adds minimal testing for the package (only activation now)
- Adds Eslint
- Using GitHub actions instead of Travis

Please add an `ATOM_ACCESS_TOKEN` for the secrets of the repo:
public-key: https://github.com/atom-ide-community/atom-ide-datatip/settings/keys 
private-key: https://github.com/atom-ide-community/atom-ide-datatip/settings/secrets

[semantic-release plugin to publish Atom packages with apm](https://github.com/semantic-release/apm#atom-authentication)

aminya committed:
chore: add test deps

aminya committed:
chore: test script

aminya committed:
chore: test runner

aminya committed:
chore: activation test

aminya committed:
chore: eslint config

aminya committed:
chore: eslint deps

aminya committed:
chore: eslint script

aminya committed:
chore: using Github actions instead of Travis

aminya commented:
If you are happy with this, we can do the same thing for other repositories. Let's get this finalized, so we can streamline other ones.

This is a successful running action: https://github.com/aminya/atom-ide-datatip/actions/runs/83546716

aminya committed:
chore: update package-lock.json

aminya commented:
@UziTech do you have any idea about why the tests fail?
https://github.com/aminya/atom-ide-datatip/runs/603872829?check_suite_focus=true#step:6:48

There is no problem with running the tests offline.

UziTech commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,75 @@
+      - name: Install APM dependencies
+        run: |
+          apm ci  # uses locked module. use `apm install` for non-locked
+      - name: Run tests 👩🏾‍💻
```
You have to install the `busy-signal` package with `apm`. It works locally because you already have that package installed.

```suggestion
      - name: Install Packages
        run: |
          apm install busy-signal
      - name: Run tests 👩🏾‍💻
```

aminya replied:
Thanks for the tip! Done in https://github.com/atom-ide-community/atom-ide-datatip/pull/66/commits/408b45d5a0d257046fb11bc8e8bcbba43384038f

UziTech commented on spec/main-spec.js:
```diff
@@ -0,0 +1,26 @@
+        atom.packages.triggerDeferredActivationHooks();
+        // Activate package-deps
+        await atom.packages.activatePackage("busy-signal")
+        await atom.packages.activatePackage("atom-ide-markdown-service")
```
It looks like you will need to install `atom-ide-markdown-service` as well. Only core packages are installed by default.

aminya committed:
chore: install atom package deps

Fixes 
https://github.com/atom-ide-community/atom-ide-datatip/pull/66#discussion_r411858904 
and 
https://github.com/atom-ide-community/atom-ide-datatip/pull/66#discussion_r411857864

Co-Authored-By: @uzitech <tony@brix.ninja>

aminya commented:
@Belar I don't have write access to this repository. Could you review and merge this?

aminya commented:
@Belar Could you review and merge this?

I don't have write access to this repository. If I have write-access myself, I can conduct the development of the packages faster. I can also fix the possible bugs directly.

Belar commented:
Thanks for the PR, looks very promising :tada:, will take a look soon. 

Of course, as a member you should be able to work within the organization, will update the setup. However, reviews should still be part of the process, at least until we stabilize everything.

aminya commented:
Could you merge this then?

Yes, I agree with the review process too. My point was more about the general chore stuff and bug fix rather directly applying breaking changes.

Belar commented on .eslintrc.json:
```diff
@@ -0,0 +1,88 @@
+//      "parserOptions": {
+//        "ecmaFeatures": { "jsx": true },
+//        "ecmaVersion": 2020,
+//        "sourceType": "module"
+//      },
+//      "plugins": ["@typescript-eslint", "only-warn"],
+//      "extends": [
+//        "eslint:recommended",
+//        "plugin:@typescript-eslint/eslint-recommended",
+//        "plugin:@typescript-eslint/recommended"
+//      ],
+//      "rules": {
+//        "@typescript-eslint/explicit-function-return-type": "off",
+//        "@typescript-eslint/camelcase": "off",
+//        "@typescript-eslint/no-use-before-define": "off",
+//        "@typescript-eslint/member-delimiter-style": "off"
+//      }
+//    },
+
+      /* CoffeeScript files */
+
+//    {
+//      "files": ["**/*.coffee"],
+//      "env": {
+//        "atomtest": true,
+//        "es6": true,
+//        "node": true,
+//        "browser": true
+//      },
+//      "globals": { "atom": "readonly" },
+//      // "parser": "eslint-plugin-coffee",
+//      "parser": "eslint-plugin-coffee",
+//      "parserOptions": {
+//        "ecmaFeatures": { "jsx": true },
+//        "ecmaVersion": 2020,
+//        "sourceType": "module"
+//      },
+//      "plugins": ["coffee", "only-warn"],
+//      "extends": ["plugin:coffee/eslint-recommended"]
+//    },
```
Since we are not using this part of the config, could we remove it? We can always add it, if/when we move to TypeScript, and it's very unlikely we will switch to CoffeeScript.

aminya replied:
I made this as a template for all the repositories. Removing some commented code (which is structured very clearly) will not make much difference. We don't want to touch this file much later. If you still think we should remove this, I am OK. I have a backup offline.

Belar replied:
Technically it doesn't make a lot of difference to remove it. However, as you said, we don't want to touch it much later; which means it will most likely get outdated, and require someone to shift focus to analyse it (when looking into the config), what makes it possible source of distraction.

Belar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,79 @@
+        uses: wagoid/commitlint-github-action@v1
+      - uses: actions/setup-node@v1
+        with:
+          node-version: "13.x"
```
`v13` is already in maintenance phase and will have its "end of life" soon. Could we move to either `v12` (active LTS) or `v14` (current, the next LTS)?

aminya replied:
Fixed in https://github.com/atom-ide-community/atom-ide-datatip/pull/66/commits/ba50f756519da9a1a188bb57d248f346dea73fcb

Belar replied:
Great, could we also update it for "Lint" step?

Belar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,79 @@
+        run: npm ci
+      - name: Release 🎉
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Currently, `atom-ide-community-bot` is used (through its `GITHUB_TOKEN`) to perform tasks around a release (e.g. [message in a PR](https://github.com/atom-ide-community/atom-ide-datatip/pull/62#issuecomment-612872810)). In Actions the token is generated automatically and [it's not recommended to work around it](https://github.com/semantic-release/semantic-release/blob/0785a844fa8ac1320383452ce531898be3b01f92/docs/recipes/github-actions.md), do you know how will this behaviour change?

aminya replied:
`GitHub_TOKEN` is Github's default token for every repository. We can use our custom token too.

The benefit of using Github action instead of a bot is its flexibility. In #67 I used this feature to run `ppublish.sh` and build the code before releasing.
https://github.com/atom-ide-community/atom-ide-datatip/pull/67/files#diff-e9f950f17198d3d5e3122a44230a09b9R74

Unless `atom-ide-community-bot` allows running a custom pre-publish script, I don't think it gives any benefit over Github Actions.


Other packages using this:
https://github.com/AtomLinter/linter-coffeelint/blob/master/.github/workflows/main.yml#L45-L62
Configuration of Token:
https://github.com/semantic-release/apm#atom-authentication

Belar replied:
> Unless atom-ide-community-bot allows running a custom pre-publish script, I don't think it gives any benefit over Github Actions.

The `bot` suffix is there just to hint that's a machine user. Travis (what Actions would replace) and semantic-release are doing the actual work when it comes to releases, although they need access that comes with the bot account (through a personal token, `GITHUB_TOKEN`).

> GitHub_TOKEN is Github's default token for every repository. We can use our custom token too.

It's about permissions and what we need for a smooth running Actions with `semanatic-release`. Based on the config you linked for `AtomLinter`, and as per https://github.com/atom-ide-community/atom-ide-datatip/pull/66/files#r414364902, it will be the safest to go with `GH_TOKEN` (for the bot account's personal token) if default Action's token isn't sufficient - I don't know the current state, but saw multiple threads that it wasn't.

Belar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,79 @@
+        run: npm ci
+      - name: Release 🎉
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
As we spoke earlier, it would be nice to allow members to work within the organization (write access), but that will require protected branches (for requirements like review, passed tests etc.). Do you know how we should approach it, considering token alteration is inadvisable and with protected branches semantic release needs broader access? 

[Documentation recommends creation of a separate, dedicated secret](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token) if default token's permissions are insufficient (we can give it `write` access, if I understand docs correctly), so we may need to switch to `GH_TOKEN`, which is used by semantic release, but not pre-populated by Actions.

Belar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,79 @@
+      - name: Release 🎉
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
We are deploying to APM, shouldn't it be `ATOM_ACCESS_TOKEN` or do Actions need `NPM_TOKEN`?

aminya replied:
Fixed in https://github.com/atom-ide-community/atom-ide-datatip/pull/66/commits/999014ec065870fe4fd4896834b37d6cef6510e4

Belar commented on package.json:
```diff
@@ -38,8 +42,16 @@
     "prettier": "^1.18.2",
-    "semantic-release": "^17.0.6"
+    "semantic-release": "^17.0.6",
+    "jasmine": "^3.5.0",
+    "@types/jasmine": "^3.5.10",
+    "atom-jasmine3-test-runner": "^4.4.4",
```
What's the benefit of updating the test setup in this case? Bundled version is old, but we can check if package activates correctly with it, without extra dependencies.

aminya replied:
- Jasmine 3 is the version that is supported by the original project and community.
- Writing tests in Jasmine 1 is cumbersome and very clumsy. 
- Jasmine 3 has `await/async` which is not available in the old version. We don't need to care about the promise and other stuff anymore.
- Jasmine 3 has very good features such as `measure` that is used in this PR. https://github.com/UziTech/atom-jasmine3-test-runner
- There is no point in writing tests in the legacy version. Someone would continue to use the old version if they have written the test years ago. 

Active atom packages that use Jasmine 3: https://github.com/UziTech/atom-jasmine3-test-runner/network/dependents?package_id=UGFja2FnZS01MDQwNjI2MQ%3D%3D

Belar replied:
Thanks for details, it sounds nice. 

I took a look at the package that sets the Jasmine 3 up and the way dependencies are handled is a bit concerning, they [reside outside of the dependency tree](https://github.com/UziTech/atom-jasmine3-test-runner/blob/ba86b149764014575d4cfd2c06709db20a17a437/lib/spec-helper/spec-helper.js), being at the mercy of manual tracking. 

I don't have a good response (preferably with a suggestion), doing a pure v3 runner setup comes to mind, but would have to look into it.

Belar commented on spec/main-spec.js:
```diff
@@ -0,0 +1,26 @@
+    it("Activation", async function () {
+
+        // This makes the log visible again from the command line.
+        spyOn(console, 'log').and.callThrough();
```
Is a call through required? I believe log works without a call through

aminya replied:
Yes, generally we need it.
https://stackoverflow.com/a/54334807/7910299

However, I removed this in https://github.com/atom-ide-community/atom-ide-datatip/pull/66/commits/14984f72de2df326ba84297c8a3dd4945971e9eb

Belar commented on spec/main-spec.js:
```diff
@@ -0,0 +1,26 @@
+        // This makes the log visible again from the command line.
+        spyOn(console, 'log').and.callThrough();
+
+        console.log("\n Activation Started")
```
Debug logs? If not, how do we use this info? We know about test case's run and when it's finished (pass/fail).

aminya replied:
You just run
```
npm run test
```
in your terminal and it shows the result.


There is a graphical interface too: https://github.com/UziTech/atom-jasmine3-test-runner#atom-jasmine-3x-test-runner. However, I don't use it personally.

Belar reviewed (commented):
It would be great to have changes like this in separate PRs. For example, linting setup is just a comment removal from getting in, when for Actions it seems we may need to address a few things first (maybe even do a test release with our setup, in a test package).

Nice to see these coming in.

aminya committed:
chore: using Node 14

aminya committed:
chore: ATOM_ACCESS_TOKEN

aminya commented:
I update the OP for the correct semantic release configuration. Since I don't have access to the settings you should do it yourself.

https://github.com/semantic-release/apm#atom-authentication

aminya committed:
chore: separating benchmark from spec

aminya commented:
@Belar Could you review this again? I think the only thing left to set up is the token for semantic release (or option to run a custom prepublish script on the atom-ide-community bot)

aminya committed:
chore: use cycjimmy/semantic-release-action@v2 action

Cleans up package.json

Belar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,83 @@
+            @semantic-release/apm-config
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_TOKEN: ${{ secrets.ATOM_ACCESS_TOKEN  }}
```
Both need to use `ATOM_ACCESS_TOKEN`, otherwise it won't be picked up by `@semantic-release/apm-config`.

```suggestion
          ATOM_ACCESS_TOKEN: ${{ secrets.ATOM_ACCESS_TOKEN  }}
```

aminya replied:
Fixed in https://github.com/atom-ide-community/atom-ide-datatip/pull/66/commits/c611ada7745d4635df9630fb349b0d6932211083

Belar commented on package.json:
```diff
@@ -38,14 +40,18 @@
     "commitlint": "^8.3.5",
     "husky": "^4.2.5",
     "prettier": "^1.18.2",
-    "semantic-release": "^17.0.6"
```
Does replacing `semantic-release` package with `cycjimmy/semantic-release-action@v2` action outweigh risk associated with adding another point of trust? That's not an official `semantic-release` action, unfortunately it doesn't seem like they have one.

aminya replied:
I don't think any risk is involved. That is a tested action with support. After all, we are monitoring the releases. If there was a syntax error in my action, I can make another PR to fix that.

Belar reviewed (commented):
Went over the changes, thanks for patience.

aminya committed:
chore: remove TS/CS from Eslint config

aminya commented:
@Belar Regarding Jasmine and Semantic Release, reusing the code is the best practice in software development. Redoing everything from scratch (for the excuse of safety, etc) is not a good solution. 

This is especially the case when there are only 2-3 people working on a project. We are a small community of developers, and we should do our best to speed up the process by maximizing code reuse and minimizing the effort for maintenance. To understand what I mean: 
From the time I submitted this PR, 1105 commits have been pushed to the VSCode base. https://github.com/microsoft/vscode/graphs/commit-activity

I prefer someone to make a package that does one thing very well. Instead of me making something from scratch that highly likely I will not put the same amount of effort into maintaining it. I know when I use atom-jasmine-runner this works without me being worried about the dependencies, writing bridges between jasmine and atom, etc. Worth mentioning by using the new modern systems like Deno (replacement for Node), package-lock.json is deprecated.
